addons/mrp/models/mrp_workorder.py

```diff
@@ -650,29 +650,35 @@ def button_start(self, raise_on_invalid_state=False):
     def button_finish(self):
         date_finished = fields.Datetime.now()
         all_vals_dict = defaultdict(lambda: self.env['mrp.workorder'])
+        to_be_ended_workorders = self.env['mrp.workorder']
+        moves = (self.move_raw_ids + self.production_id.move_byproduct_ids.filtered(lambda m: (m.operation_id == self.operation_id)))
         for workorder in self:
             if workorder.state in ('done', 'cancel'):
                 continue
-            moves = (self.move_raw_ids + self.production_id.move_byproduct_ids.filtered(lambda m: m.operation_id == self.operation_id))
+            moves = moves.filtered(lambda move: not move.picked)
             for move in moves:
-                if not move.picked:
-                    if float_is_zero(workorder.production_id.qty_producing, precision_rounding=workorder.production_id.product_uom_id.rounding):
-                        qty_available = workorder.production_id.product_qty
-                    else:
-                        qty_available = workorder.production_id.qty_producing
-                    new_qty = float_round(qty_available * move.unit_factor, precision_rounding=move.product_uom.rounding)
-                    move._set_quantity_done(new_qty)
+                if float_is_zero(workorder.production_id.qty_producing, precision_rounding=workorder.production_id.product_uom_id.rounding):
+                    qty_available = workorder.production_id.product_qty
+                else:
+                    qty_available = workorder.production_id.qty_producing
+                new_qty = float_round(qty_available * move.unit_factor, precision_rounding=move.product_uom.rounding)
+                move._set_quantity_done(new_qty)
             moves.picked = True
-            workorder.end_all()
+            to_be_ended_workorders += workorder
+
+        to_be_ended_workorders.end_all()
+
+        for workorder in to_be_ended_workorders:
             vals = {
-                'qty_produced': workorder.qty_produced or workorder.qty_producing or workorder.qty_production,
-                'state': 'done',
-                'date_finished': date_finished,
-                'costs_hour': workorder.workcenter_id.costs_hour
-            }
+                    'qty_produced': workorder.qty_produced or workorder.qty_producing or workorder.qty_production,
+                    'state': 'done',
+                    'date_finished': date_finished,
+                    'costs_hour': workorder.workcenter_id.costs_hour
+                }
             if not workorder.date_start or date_finished < workorder.date_start:
                 vals['date_start'] = date_finished
             all_vals_dict[frozenset(vals.items())] |= workorder
+
```

Comment:
The changes here are formatting, which we don't do in stable versions -> pollution of history.